In [1]:
!unzip /content/STAT+S01V20.zip

Archive:  /content/STAT+S01V20.zip
  inflating: Solution_for_Project_3.ipynb  
  inflating: USArrests.xlsx          


In [2]:
import pandas as pd
import numpy as np

In [4]:
us_arrest_data = pd.read_excel("/content/USArrests.xlsx")
us_arrest_data.head()

Unnamed: 0  Murder  Assault  UrbanPop  Fraud
0     Alabama    13.2      236        58   21.2
1      Alaska    10.0      263        48   44.5
2     Arizona     8.1      294        80   31.0
3    Arkansas     8.8      190        50   19.5
4  California     9.0      276        91   40.6

**Analyze measures of spread and shape (kurtosis, skewness) on Assault and Fraud columns.**

In [16]:
#Range
print("Range of Assault column:",np.ptp(us_arrest_data['Assault']))
print("Range of Fraud column:",np.ptp(us_arrest_data['Fraud']))

#IQR
Q1_assault = us_arrest_data['Assault'].quantile(0.25)
Q3_assault = us_arrest_data['Assault'].quantile(0.75)
IQR_assault = Q3_assault - Q1_assault
print("IQR of Assault column:",IQR_assault)

Q1_fraud = us_arrest_data['Fraud'].quantile(0.25)
Q3_fraud = us_arrest_data['Fraud'].quantile(0.75)
IQR_fraud = Q3_fraud - Q1_fraud
print("IQR of Fraud column:",IQR_fraud)

#Variance
print("Vaiance of Assault column:",us_arrest_data['Assault'].var())
print("Variance of Fraud column:",us_arrest_data['Fraud'].var())

#Standard Deviation
print("Standard Deviation of Assault column:",us_arrest_data['Assault'].std())
print("Standard Deviation of Fraud column:",us_arrest_data['Fraud'].std())

Range of Assault column: 292
Range of Fraud column: 38.7
IQR of Assault column: 140.0
IQR of Fraud column: 11.100000000000001
Vaiance of Assault column: 6945.165714285714
Variance of Fraud column: 87.72915918367349
Standard Deviation of Assault column: 83.33766084001707
Standard Deviation of Fraud column: 9.36638453105965


The above output indicates that teh range of Assault column is much greater than the Fraud column and hence the data in Assault column is greatly varied and distributed than Fraud column.

In [23]:
from scipy.stats import skew, kurtosis
print("Skewness of Assault column:",skew(us_arrest_data['Assault']))
print("Skewness of Fraud column:",skew(us_arrest_data['Fraud']))

print(f"Pearson kurtosis of Assault column: {kurtosis(us_arrest_data['Assault'], fisher = False)}")
print(f"Fisher's kurtosis of Assault column: {kurtosis(us_arrest_data['Assault'], fisher = True)}")

print(f"Pearson kurtosis of Fraud column: {kurtosis(us_arrest_data['Fraud'], fisher = False)}")
print(f"Fisher's kurtosis of Fraud column: {kurtosis(us_arrest_data['Fraud'], fisher = True)}")

Skewness of Assault column: 0.2273178665911123
Skewness of Fraud column: 0.7769613202124208
Pearson kurtosis of Assault column: 1.930979948426923
Fisher's kurtosis of Assault column: -1.069020051573077
Pearson kurtosis of Fraud column: 3.201897793084265
Fisher's kurtosis of Fraud column: 0.20189779308426514


- Skewness: 0.2273178665911123
A skewness value of 0.227 indicates that the data is slightly positively skewed (right-skewed). This means there is a slight asymmetry with a longer tail on the right side of the distribution, but it is close to being symmetrical.
Fraud Column:

- Skewness: 0.7769613202124208
A skewness value of 0.777 indicates that the data is moderately positively skewed (right-skewed). This suggests a more noticeable asymmetry with a longer tail on the right side of the distribution.
Kurtosis
Assault Column:

- Pearson Kurtosis: 1.930979948426923
Fisher's Kurtosis: -1.069020051573077
Pearson kurtosis of 1.931 (which is less than 3) and Fisher's kurtosis of -1.069 (since Fisher's kurtosis is calculated as Pearson kurtosis minus 3) indicate that the distribution is platykurtic. This means the distribution has thinner tails and a flatter peak compared to a normal distribution. There are fewer extreme outliers than a normal distribution.
Fraud Column:

- Pearson Kurtosis: 3.201897793084265
Fisher's Kurtosis: 0.20189779308426514
Pearson kurtosis of 3.202 (which is slightly greater than 3) and Fisher's kurtosis of 0.202 indicate that the distribution is slightly leptokurtic. This means the distribution has slightly fatter tails and a sharper peak compared to a normal distribution. There are more extreme outliers than a normal distribution, but the difference is not very pronounced.

**Obtain z-score based on data in Assault column.**

In [24]:
from scipy.stats import zscore
print("Assault column z-score:")
print(zscore(us_arrest_data['Assault']))

Assault column z-score:
0     0.790787
1     1.118060
2     1.493817
3     0.233212
4     1.275635
5     0.402909
6    -0.736484
7     0.815030
8     1.990786
9     0.487757
10   -1.512241
11   -0.615272
12    0.948363
13   -0.700121
14   -1.391029
15   -0.675878
16   -0.748605
17    0.948363
18   -1.063757
19    1.566544
20   -0.263757
21    1.021090
22   -1.197090
23    1.069575
24    0.087757
25   -0.748605
26   -0.833454
27    0.984726
28   -1.378908
29   -0.142545
30    1.384726
31    1.008969
32    2.015028
33   -1.524362
34   -0.615272
35   -0.239515
36   -0.142545
37   -0.784969
38    0.039273
39    1.311999
40   -1.027393
41    0.208970
42    0.366545
43   -0.615272
44   -1.487999
45   -0.178909
46   -0.312242
47   -1.087999
48   -1.427393
49   -0.118303
Name: Assault, dtype: float64


In [26]:
from sklearn.preprocessing import StandardScaler
ss = StandardScaler()
ss.fit_transform(us_arrest_data['Assault'].values.reshape(-1,1))

array([[ 0.79078716],
       [ 1.11805959],
       [ 1.49381682],
       [ 0.23321191],
       [ 1.2756352 ],
       [ 0.40290872],
       [-0.73648418],
       [ 0.81502956],
       [ 1.99078607],
       [ 0.48775713],
       [-1.51224105],
       [-0.61527217],
       [ 0.94836277],
       [-0.70012057],
       [-1.39102904],
       [-0.67587817],
       [-0.74860538],
       [ 0.94836277],
       [-1.06375661],
       [ 1.56654403],
       [-0.26375734],
       [ 1.02108998],
       [-1.19708982],
       [ 1.06957478],
       [ 0.0877575 ],
       [-0.74860538],
       [-0.83345379],
       [ 0.98472638],
       [-1.37890783],
       [-0.14254532],
       [ 1.38472601],
       [ 1.00896878],
       [ 2.01502847],
       [-1.52436225],
       [-0.61527217],
       [-0.23951493],
       [-0.14254532],
       [-0.78496898],
       [ 0.03927269],
       [ 1.3119988 ],
       [-1.027393  ],
       [ 0.20896951],
       [ 0.36654512],
       [-0.61527217],
       [-1.48799864],
       [-0

**Detect outliers based on data in UrbanPop column using IQR.**

In [31]:
Q1_urbanpop = us_arrest_data['UrbanPop'].quantile(0.25)
Q3_urbanpop = us_arrest_data['UrbanPop'].quantile(0.75)
IQR_urbanpop = Q3_urbanpop - Q1_urbanpop
print("IQR of UrbanPop column is:",IQR_urbanpop)

outlier_1 = us_arrest_data['UrbanPop'] < Q1_urbanpop - 1.5*IQR_urbanpop
outlier_2 = us_arrest_data['UrbanPop'] > Q3_urbanpop + 1.5*IQR_urbanpop
print("Outliers in UrbanPop column are:",us_arrest_data['UrbanPop'][outlier_1 |outlier_2])

IQR of UrbanPop column is: 23.25
Outliers in UrbanPop column are: 20    30000
Name: UrbanPop, dtype: int64


From the above output we can conclude that the UrbanPop column has only 1 outlier i.e., row 20 and the IQR of UrbanPop is 23.25.

**Detect outliers based on UrbanPop column using z-score.**

In [34]:
from scipy.stats import zscore
print("UrbanPop column z-score:")
z = pd.DataFrame(zscore(us_arrest_data['UrbanPop']))
z

UrbanPop column z-score:


UrbanPop
0  -0.144561
1  -0.146947
2  -0.139311
3  -0.146470
4  -0.136687
5  -0.139788
6  -0.140027
7  -0.141220
8  -0.139311
9  -0.144083
10 -0.138595
11 -0.145515
12 -0.138595
13 -0.142890
14 -0.144799
15 -0.142652
16 -0.145992
17 -0.142652
18 -0.146231
19 -0.142413
20  6.999961
21 -0.140743
22 -0.142652
23 -0.147901
24 -0.141697
25 -0.145754
26 -0.143606
27 -0.139073
28 -0.145038
29 -0.137164
30 -0.141697
31 -0.137880
32 -0.147663
33 -0.147901
34 -0.140504
35 -0.142175
36 -0.142413
37 -0.141220
38 -0.137641
39 -0.146947
40 -0.147663
41 -0.144322
42 -0.139311
43 -0.139311
44 -0.150765
45 -0.143368
46 -0.140982
47 -0.149094
48 -0.142652
49 -0.144083

In [35]:
print(np.where(z > 1))

(array([20]), array([0]))


In the above code cell, we have mentioned a specific threshold value for detecting outliers i.e., if the zscore value is gretaer than 1, it'll be considered as an outlier. Hence, row 20 has a zscore value greater than 1 and is an outlier.

**Calculate Range, variance, and Standard Deviation of data in UrbanPop and Assault column.**

In [46]:
print("Range of UrbanPop column:",np.ptp(us_arrest_data['UrbanPop']))
print("Range of Assault column:",np.ptp(us_arrest_data['Assault']))
print("Variance of UrbanPop column:",us_arrest_data['UrbanPop'].var())
print("Variance of Assault column:",us_arrest_data['Assault'].var())
print("Standard Deviation of UrbanPop column:",us_arrest_data['UrbanPop'].std())
print("Standard Deviation of Assault column:",us_arrest_data['Assault'].std())

Range of UrbanPop column: 29968
Range of Assault column: 292
Variance of UrbanPop column: 17922115.07591836
Variance of Assault column: 6945.165714285714
Standard Deviation of UrbanPop column: 4233.451910193189
Standard Deviation of Assault column: 83.33766084001707


From the above output we can conclude that UrbanPop column has more range than Assault column. The data in UrbanPop is also much more varied and spread out than the data in Assault column.